### ISSUE TITLE
Linker error when building with GCC 10

### ISSUE THREAD

**Issue by benide** [akermu/emacs-libvterm#296]
When trying to compile libvterm, everything works up until linking. It looks like `elisp.c` and `vterm-module.c` both include `elisp.h`, and maybe this is a problem? I don't really know what I'm talking about here, but this is as far as I could get with trying to understand the error. I've included the commands I ran with all output below:

```
[ben: ~/.config/emacs/straight/build/vterm] $ mkdir -p build
[ben: ~/.config/emacs/straight/build/vterm] $ cd build
[ben: ~/.config/emacs/straight/build/vterm/build] $ cmake ..
-- The C compiler identification is GNU 10.0.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- No build type selected, defaulting to RelWithDebInfo
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ben/.config/emacs/straight/build/vterm/build
[ben: ~/.config/emacs/straight/build/vterm/build] $ make
Scanning dependencies of target libvterm
[  8%] Creating directories for 'libvterm'
[ 16%] Performing download step (git clone) for 'libvterm'
Cloning into 'libvterm'...
Note: switching to '65dbda3ed214f036ee799d18b2e693a833a0e591'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 65dbda3 Bump version to 0.1.3
[ 25%] No patch step for 'libvterm'
[ 33%] Performing update step for 'libvterm'
[ 41%] No configure step for 'libvterm'
[ 50%] Performing build step for 'libvterm'
TBL src/encoding/DECdrawing.tbl
TBL src/encoding/uk.tbl
CC src/encoding.c
CC src/keyboard.c
CC src/mouse.c
CC src/parser.c
CC src/pen.c
CC src/screen.c
CC src/state.c
CC src/unicode.c
src/unicode.c:306:12: warning: ‘mk_wcswidth_cjk’ defined but not used [-Wunused-function]
  306 | static int mk_wcswidth_cjk(const uint32_t *pwcs, size_t n)
      |            ^~~~~~~~~~~~~~~
src/unicode.c:215:12: warning: ‘mk_wcswidth’ defined but not used [-Wunused-function]
  215 | static int mk_wcswidth(const uint32_t *pwcs, size_t n)
      |            ^~~~~~~~~~~
CC src/vterm.c
LINK libvterm.la
CC bin/unterm.c
CC bin/vterm-ctrl.c
CC bin/vterm-dump.c
[ 58%] No install step for 'libvterm'
[ 66%] Completed 'libvterm'
[ 66%] Built target libvterm
Scanning dependencies of target vterm-module
[ 75%] Building C object CMakeFiles/vterm-module.dir/vterm-module.c.o
[ 83%] Building C object CMakeFiles/vterm-module.dir/utf8.c.o
[ 91%] Building C object CMakeFiles/vterm-module.dir/elisp.c.o
[100%] Linking C shared module ../vterm-module.so
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:32: multiple definition of `Fsymbol_value'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:32: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:33: multiple definition of `Flength'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:33: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:34: multiple definition of `Flist'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:34: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:35: multiple definition of `Fnth'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:35: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:48: multiple definition of `Fput_text_property'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:48: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:49: multiple definition of `Fadd_text_properties'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:49: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:36: multiple definition of `Ferase_buffer'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:36: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:37: multiple definition of `Finsert'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:37: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:38: multiple definition of `Fgoto_char'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:38: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:40: multiple definition of `Fforward_line'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:40: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:41: multiple definition of `Fgoto_line'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:41: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:8: multiple definition of `Qt'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:8: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:42: multiple definition of `Fdelete_lines'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:42: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:9: multiple definition of `Qnil'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:9: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:43: multiple definition of `Frecenter'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:43: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:46: multiple definition of `Fpoint'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:46: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:44: multiple definition of `Fset_window_point'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:44: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:45: multiple definition of `Fwindow_body_height'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:45: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:57: multiple definition of `Feq'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:57: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:39: multiple definition of `Fforward_char'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:39: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:52: multiple definition of `Fget_buffer_window_list'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:52: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:53: multiple definition of `Fselected_window'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:53: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:25: multiple definition of `Qcursor_type'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:25: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:50: multiple definition of `Fset'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:50: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:58: multiple definition of `Fvterm_get_color'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:58: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:54: multiple definition of `Fvterm_set_title'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:54: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:55: multiple definition of `Fvterm_set_directory'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:55: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:56: multiple definition of `Fvterm_invalidate'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:56: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:59: multiple definition of `Fvterm_eval'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:59: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:51: multiple definition of `Fvterm_flush_output'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:51: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:29: multiple definition of `Qvterm_prompt'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:29: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:28: multiple definition of `Qrear_nonsticky'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:28: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:27: multiple definition of `Qvterm_line_wrap'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:27: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:26: multiple definition of `Qemacs_major_version'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:26: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:24: multiple definition of `Qhbar'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:24: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:23: multiple definition of `Qbar'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:23: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:22: multiple definition of `Qbox'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:22: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:21: multiple definition of `Qface'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:21: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:20: multiple definition of `Qextend'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:20: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:19: multiple definition of `Qstrike'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:19: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:18: multiple definition of `Qreverse'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:18: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:17: multiple definition of `Qslant'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:17: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:16: multiple definition of `Qunderline'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:16: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:15: multiple definition of `Qweight'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:15: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:14: multiple definition of `Qbackground'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:14: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:13: multiple definition of `Qforeground'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:13: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:12: multiple definition of `Qitalic'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:12: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:11: multiple definition of `Qbold'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:11: first defined here
/usr/bin/ld: CMakeFiles/vterm-module.dir/elisp.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:10: multiple definition of `Qnormal'; CMakeFiles/vterm-module.dir/vterm-module.c.o:/home/ben/.config/emacs/straight/build/vterm/elisp.h:10: first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/vterm-module.dir/build.make:135: ../vterm-module.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:153: CMakeFiles/vterm-module.dir/all] Error 2
make: *** [Makefile:104: all] Error 2
```

**Comment by benide** [comment]
I did a little digging and figured out that it's because I'm using GCC 10, which defaults to using the flag `-fno-common`, and indeed this does cause problems with including the same header twice where variables are not declared `extern`. See [here](https://gcc.gnu.org/gcc-10/porting_to.html).

Simply adding `extern` to all of the variables in `elisp.h` doesn't work because then they are declared, but never defined which causes runtime errors, so it's a little deeper than I (who knows no C except what I've just googled today) can do anything with.

**Comment by Sbozzolo** [comment]
Thanks for the bug report. For the moment, you can compile with the `-fcommon` flag. We will go to the root of the problem soon.

**Comment by kirill-gerasimenko** [comment]
Hey! 

@Sbozzolo how do I do what you suggested? I tried adding this in the cmake txt file to CFLAGS but it won't make any difference.. 

Thanks!

**Comment by nikolauska** [comment]
@kirill-gerasimenko It worked for me when setting `COMPILE_FLAGS` in the `set_target_properties` for the `vterm-module`

Example:
```
set_target_properties(vterm-module PROPERTIES
  C_STANDARD 99
  POSITION_INDEPENDENT_CODE ON
  PREFIX ""
  LIBRARY_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}
  COMPILE_FLAGS "-fcommon"
  )

```

**Comment by kirill-gerasimenko** [comment]
@nikolauska that's worked, thanks a lot!

**Comment by Sbozzolo** [comment]
Fixed in #319
